### ISSUE TITLE
Bugs with `workspace/didChangeWatchedFiles` handling

### ISSUE THREAD

**Issue by KostkaBrukowa** [neovim/neovim#24231]
### Problem

Hey! I'm working on integrating `workspace/didChangeWatchedFiles` in `typescript-tools.nvim` and I've found some issues with that:
1. When a file is deleted, wrong update type is sent to the LS. According to LSP spec:

<details>

```
export namespace [WatchKind](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#watchKind) {
	/**
	 * Interested in create events.
	 */
	export const Create = 1;

	/**
	 * Interested in change events
	 */
	export const Change = 2;

	/**
	 * Interested in delete events
	 */
	export const Delete = 4;
}
```

</details>

event type for `Delete` should be `4` but [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/_watch.lua#L7) on `Delete` event is equal to `3`

2. Also wrong event type is sent after renaming a file. When creating new file params sent to the server are:
```
-- creating `main4.tsx`
{
  changes = { {
      type = 1,
      uri = "file:///Users/jaroslaw.glegola/Documents/Praca/node-typescript-boilerplate-main/src/main4.tsx"
    } }
}
```
but after rename, `type` is also the same
```
-- renaming to `main5.tsx`
{
  changes = { {
      type = 1,
      uri = "file:///Users/jaroslaw.glegola/Documents/Praca/node-typescript-boilerplate-main/src/main5.tsx"
    } }
}
```
It could be, because [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/_watch.lua#L66) when we have `events.rename` response, we are sending wrong type.

3. Using `parse` function from [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/_watchfiles.lua#L14) and `_match` from [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/_watchfiles.lua#L83) and using this code 
```
  local reg = parse("**/*.{ts,tsx}")

  print(M._match(reg, "/test.tsx"))
```
printed result is `false`. Surprisingly
```
  local reg = parse("**/*.{tsx,ts}")

  print(M._match(reg, "/test.tsx"))
```
works fine and printed result is `true`

### Steps to reproduce

If some branch is needed I can send one from `typescript-tools` but basically going into the code and putting some `print` statements should be enough.

### Expected behavior

1. Delete event type should be `4`
2. After rename event type should be `2`
3. Globs should correctly match paths

### Neovim version (nvim -v)

NVIM v0.10.0-dev-565+g3688735c2-dirty

### Vim (not Nvim) behaves the same?

not related

### Operating system/version

13.3.1 (a)

### Terminal name/version

kitty Version 0.26.4 (0.26.4)

### $TERM environment variable

truecolor

### Installation

homebrew

**Comment by KostkaBrukowa** [comment]
To clarify second point.
If e.g. server starts with no files tracking, when we `Create` first file `test1.ts` the list of tracked files are
```
test1.ts
```
then after renaming `test1.ts` nvim should send two events, because it only sends one: `Create` event, and after renaming `test1.ts` -> `test2.ts` tracked files would be:
```
test1.ts
test2.ts
```
instead of only one file

**Comment by lewis6991** [comment]
This is not API that's available for public use.

**Comment by KostkaBrukowa** [comment]
Ok then, so let me rephrase that:
When I dynamically register capability using this piece of code:
```lua
dispatchers.server_request("client/registerCapability", {
      registrations = {
        {
          id = "123-123-123-123",
          method = "workspace/didChangeWatchedFiles",
          registerOptions = {
            watchers = {
              {
                globPattern = "**/*.{ts,tsx}",
              },
            },
          },
        },
      },
    })
```
1. There is no event sent to the server when I rename file e.g. from `test.tsx -> test1.tsx`
2. There is no delete event after I rename a file

**Comment by justinmk** [comment]
@nojnhuh

**Comment by nojnhuh** [comment]
Neovim uses LPeg's `+` ordered-choice operator for `{}` conditions, and since `+` doesn't backtrack once the pattern fails after matching `ts`, the whole pattern fails without trying `tsx`, so reversing the options in the pattern is a sensible workaround here. That kind of backtracking seems necessary here to handle this case, but I have a feeling building that on top of LPeg would turn pretty gnarly.

As for the particulars around what events get emitted from the non-polling watcher, that's all consumed directly from libuv, so I'm not sure how much Neovim can reasonably massage those events to do exactly what you expect. From what I've seen, LSP servers don't rely on this kind of data to be precise enough to maintain a perfectly consistent view of the filesystem and instead mostly interpret these events at the level of "something changed around here" and check the actual state of the filesystem directly as needed. The Nodejs docs provide a number of caveats when using libuv to watch files which may be relevant to what you're seeing: https://nodejs.org/api/fs.html#caveats
